### PR TITLE
feat(redpanda): Add option for arbitrary bootstrap config

### PR DESCRIPTION
**Pull request by oleiman**
## What does this PR do?

WithBootstrapConfig adds an arbitrary config kvp to the Redpanda container. Per the name, this config will be interpolated into the generated bootstrap config file, which is particularly useful for configs requiring a restart when otherwise applied to a running Redpanda instance.

## Why is it important?

Several Redpanda cluster config properties can be updated live but require a node restart to take effect. The simplest way to expose these to testcontainer consumers is to provide generic access prior to container startup.

An example usage case is an integration test for a particularly memory hungry Wasm Data Transform, where tuning the per core and per function memory limits in a flexible, implementation driven way may be the difference between testing or not testing that function.

## Related issues

-

### CODE REVIEW

**Review by mdelapenya — COMMENTED** [review]
Could you add tests for this new option? 

Once there, I think this is LGTM 

Cheers!

**Comment by oleiman** [comment]
> Could you add tests for this new option? 
> 
> Once there, I think this is LGTM 
> 
> Cheers!

Absolutely. Thanks for the quick review @mdelapenya !

**Review by kiview — APPROVED** [review]

**Comment by oleiman** [comment]
[force push](https://github.com/testcontainers/testcontainers-go/compare/c69723c2623e000eb23083414224c725f99dce45..86985f06fabbeeb4acee69865e2e3aa23cb16e2b) contents: a test

**Comment by oleiman** [comment]
Hey @mdelapenya - does the test I added look ok to you?

**Comment by rockwotj** [comment]
Ping @mdelapenya - I think the concerns here are addressed - thanks in advance!

**Comment by mdelapenya** [comment]
I'm on PTO for this week, will take a look then 🙏

File: modules/redpanda/mounts/bootstrap.yaml.tpl
Comment (stevenh) [inline modules/redpanda/mounts/bootstrap.yaml.tpl:30]: suggestion: I don't believe you need the if here as zero iterations should be fine?
```suggestion
{{- range $key, $value := .ExtraBootstrapConfig }}
{{ $key }}: {{ $value }}
{{- end }}
```

File: modules/redpanda/options.go
Comment (stevenh) [inline modules/redpanda/options.go:164]: question: This seems like it might get cumbersome to specify multiple values, is there a reason we don't just allow a string to include, which would support arbitrary levels of yaml too?

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:610]: bug: As noted by the linter some of these parameters are reversed.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:599]: question: Is there a need for extra indentation?

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:613]: bug: close needs to be directly after the error check for the request otherwise it may not be called.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:629]: bug: close needs to be directly after the error check for the request otherwise it may not be called.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:605]: suggestion: use `require` instead of assert, as if this fails everything else below will as well, but it will just be unhelpful.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:622]: suggestion: use `require` instead of assert, as if this fails everything else below will as well, but it will just be unhelpful more below.

**Review by stevenh — CHANGES_REQUESTED** [review]
Quick pass to help keep this moving along, some issues and questions.

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go:610]: > as noted by the linter
ah, totally missed that. thanks!

**Comment by oleiman** [comment]
[force push](https://github.com/testcontainers/testcontainers-go/compare/86985f06fabbeeb4acee69865e2e3aa23cb16e2b..cc28e932195b43113dabc2eef009f68f57171f1d) some CR feedback:
- linter errors
- template cruft
- better testify usage

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go]: not sure I understand the bug here, but I've moved `close` closer to the last use of `resp.Body`.

File: modules/redpanda/options.go
Comment (mdelapenya) [reply modules/redpanda/options.go:164]: I like having a map because readers will identify the key-value pairs very easily. Of course you can do it with a multiline string, but I see using a map more "typed". OTOH if you all agree the string is fine, I'm fine too

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [reply modules/redpanda/redpanda_test.go]: To clarify, we always want the body to be closed so if you want to leverage defer:
```go
res, err := http.Get(requestURL)
require.NoError(t, err)
defer resp.Body.Close()

// Decode body and check for errors.
```

File: modules/redpanda/options.go
Comment (stevenh) [reply modules/redpanda/options.go:164]: The reason I asked was yaml is typically structured, so not all values are at the base level like the example, [here's full config example](https://docs.redpanda.com/current/reference/console/config/) so if if none route elements might be needed this approach would be limited.

Someone who knows redpanda should comment if structured config for bootstrap could be needed?

Comment (rockwotj) [reply modules/redpanda/options.go:164]: Our cluster config and thus bootstrap yaml file is always a flat key value pair - nesting is not allowed

Comment (stevenh) [reply modules/redpanda/options.go:164]: Thanks for confirming

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:609]: To clarify my previous comment about the location of `Body.Close()`, here's an example which ensures that it's always closed.

While its not so important for tests, its a really good habit to get into as it ensure that underlying transport connections can be reused, hope that helps.
```suggestion
		require.NoError(t, err)
		defer resp.Body.Close()
		
		require.Equal(t, http.StatusOK, resp.StatusCode)
		var data map[string]interface{}
		err = json.NewDecoder(resp.Body).Decode(&data)
		require.NoError(t, err)
```

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:626]: See above for how to ensure that this is always closed.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:606]: nit: we can use any here.
```suggestion
		var data map[string]any
```

**Review by stevenh — CHANGES_REQUESTED** [review]
Thanks for the previous changes, have clarified my previous commend about Body.Close and provided an example, so hope that helps.

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go]: thanks for clarifying

**Comment by oleiman** [comment]
[force push](https://github.com/testcontainers/testcontainers-go/compare/cc28e932195b43113dabc2eef009f68f57171f1d..5899844fc0bb1015fb46621c53b4e4723f0a99b0) contents:
- correct response body closure in tests
- `s/interface{}/any/`

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go:599]: not an outright need, just limiting identifier scope for my own convenience.

**Comment by oleiman** [comment]
Thanks for the reviews @stevenh - I think we've addressed all of your comments; please lmk if you have any lingering concerns 🙂

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:611]: Still a few asserts.

File: modules/redpanda/redpanda_test.go
Comment (stevenh) [inline modules/redpanda/redpanda_test.go:627]: suggestion: This is actually a good example of why require is preferable to assert.

If the length here was zero the assert would mark the test as failed, but would allow it continue, the following line would then panic as it indexed into an empty slice.

**Review by stevenh — CHANGES_REQUESTED** [review]
Just a few asserts remaining if you would be so kind.

As you can see from the example, they aren't an issue until a test actually fails, then they can result it quite hard to understand output as you typically end up with compound errors.

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go:627]: good catch, yeah. need to have it beaten into my brain that `assert` is the non-halting version...

File: modules/redpanda/redpanda_test.go
Comment (oleiman) [reply modules/redpanda/redpanda_test.go:611]: I don't see a reason to halt here since the error is non-fatal with respect to subsequent checks, but I'll happily to defer to idiomatic usage.

**Review by mdelapenya — APPROVED** [review]
LGTM, thanks all for your work here 🙇

**Review by stevenh — APPROVED** [review]

File: modules/redpanda/mounts/bootstrap.yaml.tpl
Comment (mdelapenya) [inline modules/redpanda/mounts/bootstrap.yaml.tpl:26]: Question: how will redpanda behave if users pass incorrect config key-value pairs here?

This is not related to the code itself but to the experience using the module

**Comment by oleiman** [comment]
[rebase main](https://github.com/testcontainers/testcontainers-go/compare/0e1db02a08ebaedb7c180d0b031e4f39aadf383a..98389fe8efa5ba473a2bb194a6cf6cb68a975a57)

File: modules/redpanda/mounts/bootstrap.yaml.tpl
Comment (oleiman) [reply modules/redpanda/mounts/bootstrap.yaml.tpl:26]: Sorry, missed this. In general, redpanda will simply reject any unrecognized config (the key) and halt. I'm fairly certain the same applies to ill-formed values. For example, a value outside the range of a bounded config. Broker logs should reflect this in detail.

So the experience configuring the module this way should be quite similar to editing a Redpanda cluster config file before firing up Redpanda manually.

Comment (mdelapenya) [reply modules/redpanda/mounts/bootstrap.yaml.tpl:26]: Thanks for the explanation. Once the CI passes, I'll merge this PR

**Review by mdelapenya — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
